src/librbd/object_map/DiffRequest.cc

```diff
@@ -219,10 +219,11 @@ void DiffRequest<I>::handle_load_object_map(int r) {
   auto end_it = m_object_map.end();
   if (m_object_map.size() > prev_object_diff_state_size) {
     for (; it != end_it; ++it,++diff_it, ++i) {
-      uint8_t object_map_state = *it;
+      const uint8_t object_map_state = *it;
       if (object_map_state == OBJECT_NONEXISTENT) {
         *diff_it = DIFF_STATE_HOLE;
-      } else if (diff_from_start || object_map_state != OBJECT_EXISTS_CLEAN) {
+      } else if ((object_map_state != OBJECT_EXISTS_CLEAN) &&
+                 (diff_from_start ||  m_object_diff_state_valid)) {
```

Comment:
Reading the message log:

> While computing resize diff during object map do not mark state as new data unless diff state is valid.

Shouldn't the condition be `diff_from_start || (m_object_diff_state_valid && object_map_state != OBJECT_EXISTS_CLEAN)`?

Actually, could you please elaborate what is going on here in your opinion?

Reply:
When calculating diff between two object map we need to identify states (new data, new hole, updated data, updated hole). IMO we were missing to check if object map state is valid or not.